subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/TestWorker.java

```diff
@@ -72,7 +89,9 @@ public void execute(final WorkerProcessContext workerProcessContext) {
 
         try {
             try {
-                completed.await();
+                while (state != State.STOPPED) {
```

Comment:
Mutable state is being read unsynchronized here in the "Test worker" thread. It is changed only by the runnables which are called on the line below. That is fine, but if anyone adds an assertion or mutation in the `startProcessing`, `processTestClass`, or `stop` methods which is outside of the Runnable, those assertions will fail sporadically because those methods are called from a different thread.

This code works great as-is, but I think a command-queue which encodes the control flow without a mutable state variable will be easier to maintain, e.g. [868dee1](https://github.com/gradle/gradle/blob/868dee1594e987dfc82733d0191e70c9458a6dc8/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/TestWorker.java#L91-L120)

Reply:
Yes, this does assume that the main thread is doing all of the work synchronously and therefore synchronization is not necessary for the state object.  This was a conscious trade-off in favor of keeping it very simple to understand what happens when each method of the remote protocol is called (rather than having a slightly more complex main method).  I don't feel strongly about it, but I'm inclined to leave this as is for now and just call out this assumption explicitly in the javadoc.

Reply:
Sounds good on all fronts. Fwiw, I think it's fine as-is and doesn't need further documentation. If anybody makes changes in the future and bumps into synchronization problems it's hard to catch their attention from the docs anyway.